### PR TITLE
feat: agent CLI introspection (simplified)

### PR DESCRIPTION
**Pull request by johnnygreco**
## Summary

Adds a `data-designer agent` CLI subgroup that lets AI coding agents dynamically discover Data Designer's configuration types, builder API, and local state. All output is human-readable text.

Also adds `schema_text()` to `ConfigBase` so every config type can describe its own fields, and normalizes validator/constraint discriminators to kebab-case.

## Agent commands

| Command | Description |
|---|---|
| `agent context` | Bootstrap payload combining types, local state, and builder API — recommended starting point |
| `agent types [FAMILY]` | Type names and import paths for one or all families (`columns`, `samplers`, `validators`, `processors`, `constraints`) |
| `agent schema <FAMILY> <TYPE>` | Human-readable field summary and JSON schema for a single type |
| `agent schema <FAMILY> --all` | Schemas for every type in a family |
| `agent builder` | `DataDesignerConfigBuilder` method signatures and docstrings |
| `agent state model-aliases` | Configured model aliases, providers, and usability status |
| `agent state persona-datasets` | Persona locales, dataset sizes, and installation status |

## Design

- Commands call plain functions in `agent_introspection.py` directly — no controller or service layer
- Text formatting is separated into `agent_text_formatter.py`
- `schema_text()` on `ConfigBase` (25 lines) replaces the 299-line `pydantic_inspector` from the previous approach — schema knowledge lives on the types themselves
- All commands are lazy-loaded so `--help` stays fast (`main.py` has zero new imports)

## Comparison with `johnny/feat/agent-context`

This branch delivers the same agent CLI functionality with significantly less complexity:

| Metric | `agent-context` | This branch | Reduction |
|---|---|---|---|
| Source files touched | 22 | 9 | 59% |
| Source lines added | 1,753 | 805 | 54% |
| Test lines added | 2,374 | 635 | 73% |
| Total diff | +4,127 / -130 | +1,440 / -49 | 65% |
| Architectural layers | 3 (command → controller → service) | 1 (command → plain functions) | — |

Key differences:
- **`schema_text()` on ConfigBase** (25 lines) replaces `pydantic_inspector.py` (299 lines)
- **No controller or service layer** — commands call plain functions directly
- **Human-readable text as the only output mode** — simpler interface, no `--json`/`--compact` flags
- **Lazy loading preserved** — `main.py` has zero new imports; `--help` path stays fast

### CODE REVIEW

File: packages/data-designer-config/src/data_designer/config/base.py [hunk 1]
```diff
@@ -20,6 +23,27 @@ class ConfigBase(BaseModel):
...
+        lines.append("")
+        for name, field_info in cls.model_fields.items():
+            annotation = _format_annotation(field_info.annotation)
+            if field_info.is_required():
+                lines.append(f"  {name}: {annotation}  [required]")
+            else:
+                default = field_info.default
+                if isinstance(default, Enum):
+                    default = default.value
+                lines.append(f"  {name}: {annotation} = {default!r}")
```
[the remark this answers is not in the document]
Comment (johnnygreco) [reply packages/data-designer-config/src/data_designer/config/base.py]: Fixed in a6214a98. `schema_text()` now checks `field_info.default_factory` before falling back to `field_info.default`, rendering e.g. `list()` / `dict()` instead of `PydanticUndefined`. Added 15 edge-case tests covering default_factory, enum defaults, None defaults, scalar defaults, descriptions, and docstrings.

File: packages/data-designer/src/data_designer/cli/utils/agent_introspection.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+            "default_provider": None if provider_registry is None else provider_registry.default,
+            "items": items,
+        }
+
+    providers_by_name: dict[str, Any] = {}
+    missing_key_names: set[str] = set()
+    default_provider: str | None = None
+    if provider_registry is not None:
+        providers_by_name = {p.name: p for p in provider_registry.providers}
+        default_provider = provider_registry.default or provider_registry.providers[0].name
```
[the remark this answers is not in the document]
Comment (johnnygreco) [reply packages/data-designer/src/data_designer/cli/utils/agent_introspection.py]: Fixed in a6214a98. Added a guard so an empty `providers` list falls back to `None` instead of raising `IndexError`.

File: packages/data-designer/src/data_designer/cli/utils/agent_text_formatter.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        "",
+        "## Persona Datasets",
+        "",
+        format_persona_datasets_text(data["state"]["persona_datasets"]),
+        "",
+        "## Commands",
+        "",
+        _format_table(data["operations"], ["command_pattern", "description"]),
+    ]
+    return "\n".join(sections)
```
[the remark this answers is not in the document]
Comment (johnnygreco) [reply packages/data-designer/src/data_designer/cli/utils/agent_text_formatter.py:41]: Fixed in 80b88b00. `format_context_text` now renders a `## Builder` section with the full method surface.

File: packages/data-designer/src/data_designer/cli/utils/agent_text_formatter.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    if "families" in data:
+        lines: list[str] = [f"{f['family']}: {f['count']} types" for f in data["families"]]
+        lines.append("")
+        for family_name, items in data["items"].items():
+            lines.append(_format_table(items, ["type_name", "class_name"], title=f"{family_name} types"))
+            lines.append("")
+        return "\n".join(lines).rstrip()
+    return _format_table(
+        data["items"], ["type_name", "class_name"], title=f"{data.get('family')} types" if data.get("family") else None
+    )
```
[the remark this answers is not in the document]
Comment (johnnygreco) [reply packages/data-designer/src/data_designer/cli/utils/agent_text_formatter.py:57]: Fixed in 80b88b00. Both code paths in `format_types_text` now include `import_path` in the table columns.

File: packages/data-designer/tests/cli/commands/test_agent_command.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+def test_error_outputs_message_to_stderr() -> None:
+    runner = CliRunner()
+    with patch(f"{_PATCH}.get_schema", side_effect=ValueError("boom")):
+        result = runner.invoke(app, ["agent", "schema", "columns", "missing"])
+
+    assert result.exit_code == 1
+    assert result.stdout == ""
+    assert "internal_error" in result.stderr
+    assert "boom" in result.stderr
```
[the remark this answers is not in the document]
Comment (johnnygreco) [reply packages/data-designer/tests/cli/commands/test_agent_command.py:65]: This is a false positive. Typer's `CliRunner` (which we use here) doesn't have a `mix_stderr` parameter — it captures stderr separately by default, unlike Click's `CliRunner`. The test passes as-is and `result.stderr` works correctly.

File: packages/data-designer/src/data_designer/cli/utils/agent_text_formatter.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    column_labels: dict[str, str] | None = None,
+) -> str:
+    labels = {col: (column_labels or {}).get(col, col) for col in columns}
+
+    if not items:
+        header = f"# {title}" if title else "# table"
+        return f"{header}\n(no items)"
+
+    col_widths = {col: max(len(labels[col]), max(len(str(row.get(col, ""))) for row in items)) for col in columns}
+
```
[the remark this answers is not in the document]
Comment (johnnygreco) [reply packages/data-designer/src/data_designer/cli/utils/agent_text_formatter.py:120]: Fixed in d41d4734. Both the width calculation and row rendering now use `row.get(col) or ''` to coerce `None` to empty string.

File: packages/data-designer/tests/cli/utils/test_agent_introspection_integration.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+from pathlib import Path
+
+from data_designer.cli.repositories.model_repository import ModelConfigRegistry, ModelRepository
+from data_designer.cli.repositories.provider_repository import ModelProviderRegistry, ProviderRepository
+from data_designer.cli.utils.agent_introspection import (
+    get_context,
+    get_model_aliases_state,
+    get_persona_datasets_state,
+)
```
[the remark this answers is not in the document]
Comment (johnnygreco) [reply packages/data-designer/tests/cli/utils/test_agent_introspection_integration.py:14]: Fixed in defcd284. Moved to `tests/cli/utils/test_agent_introspection_integration.py` — consistent with the flat architecture.

**Review by nabinchha — APPROVED** [review]
**Summary**: Adds a `data-designer agent` CLI subgroup that lets AI coding agents dynamically discover Data Designer's configuration types, builder API, and local state via human-readable text output. Also adds `schema_text()` to `ConfigBase` and normalizes validator/constraint discriminators to concrete `Literal` fields. The implementation matches the stated intent well — the simplified single-layer architecture (commands call plain functions directly) is a significant improvement over the prior 3-layer approach.

## Findings

### Critical — Must fix before merge

No critical issues found.

### Warnings — Strongly recommend fixing

**`packages/data-designer/src/data_designer/cli/utils/agent_text_formatter.py:119,128` — `or ''` coercion suppresses falsy values like `0` and `False`**
- **What**: `row.get(col) or ''` coerces `None` to empty string (good), but also coerces `0`, `False`, and empty string `""` to empty string. The `usable` column in model aliases contains boolean `True`/`False` values. When `usable` is `False`, `row.get("usable") or ''` evaluates to `''`, so the table would show an empty cell instead of `False`.
- **Why**: An agent reading the model-aliases table would see an empty `usable` column for unusable models, losing a key diagnostic signal.
- **Suggestion**: Use an explicit `None` check:

```python
def _cell_value(val: Any) -> str:
    return "" if val is None else str(val)
```

Then use `_cell_value(row.get(col))` in both the width calculation and row rendering.

**`packages/data-designer/src/data_designer/cli/main.py:110` — Implicit `and`/`or` precedence in filter expression**
- **What**: The condition `if prefix == "" and "." not in cmd.name or cmd.name.startswith(f"{prefix}.")` relies on `and` binding tighter than `or`. While logically correct, it reads as ambiguous.
- **Why**: A future maintainer adding a new command pattern could easily misread the intent and introduce a routing bug.
- **Suggestion**: Add explicit parentheses:

```python
if (prefix == "" and "." not in cmd.name) or (prefix != "" and cmd.name.startswith(f"{prefix}."))
```

**`packages/data-designer-config/src/data_designer/config/sampler_constraints.py:27-29` — `Constraint` base class is no longer abstract but not meant to be instantiated**
- **What**: The base `Constraint` class was previously `ABC` but is now a plain `ConfigBase` subclass. It declares `constraint_type: ConstraintType` as required (no default), so direct instantiation would fail with a confusing validation error.
- **Why**: If someone instantiates `Constraint` directly, the error won't clearly indicate they should use a subclass.
- **Suggestion**: Restore `ABC` on `Constraint` — it was abstract before and still shouldn't be instantiated directly.

**`packages/data-designer/tests/cli/commands/test_agent_command.py:57-65` — `test_error_outputs_message_to_stderr` may fail depending on Typer version**
- **What**: The test accesses `result.stdout` and `result.stderr` separately, but Typer's `CliRunner` behavior around stderr separation depends on the Typer/Click version pinned.
- **Why**: A test that errors instead of testing the intended behavior provides false confidence.
- **Suggestion**: Confirm this test passes in CI. If it doesn't, use `CliRunner(mix_stderr=False)` explicitly.

### Suggestions — Consider improving

**`packages/data-designer-config/tests/config/test_schema_text.py:149` — Single-letter loop variable `l`**
- **What**: `next(i for i, l in enumerate(lines) if "name: str" in l)` uses `l` which is visually similar to `1`.
- **Suggestion**: Rename to `line`.

**`packages/data-designer/tests/cli/utils/test_agent_text_formatter.py:219` — Import inside test function**
- **What**: `test_format_schema_text_on_real_config_models` imports `get_family_schema` inside the function body rather than at module level.
- **Suggestion**: Move the import to the top of the file per AGENTS.md guidelines.

## What Looks Good

- **`schema_text()` on `ConfigBase`** is an elegant 25-line solution replacing 299 lines of `pydantic_inspector`. Putting schema knowledge on the types themselves is the right design — discoverable, testable, and naturally stays in sync as fields change.

- **Clean data/presentation separation**: `agent_introspection.py` returns plain dicts, `agent_text_formatter.py` renders them. Both are independently testable and leave the door open for adding JSON output later with minimal effort.

- **Thorough edge-case testing for `schema_text()`**: The 15 tests in `test_schema_text.py` cover `default_factory`, enum defaults, `None` defaults, scalar defaults, descriptions, docstrings, and the mixed-model integration case. The `PydanticUndefined` regression test is particularly valuable.

## Verdict

**Ship it (with nits)** — The core design is sound, the implementation is clean, and the test coverage is good. The `or ''` falsy-value suppression is the most impactful issue to fix before merge since it silently drops `False` values from the model-aliases table. The other warnings are lower priority but worth addressing. Linter passes clean on all changed files.

**Review by nabinchha — APPROVED** [review]

File: packages/data-designer-config/src/data_designer/config/base.py [hunk 8]
```diff
@@ -83,3 +111,25 @@ class ProcessorConfig(ConfigBase, ABC):
...
+
+
+def _get_docstring_summary(docstring: str | None) -> str | None:
+    """Extract the first paragraph of a docstring, up to the Attributes section."""
+    if not docstring:
+        return None
+    lines: list[str] = []
+    for line in docstring.strip().splitlines():
+        stripped = line.strip()
+        if stripped.lower().startswith("attributes:"):
```
[the remark this answers is not in the document]
Comment (johnnygreco) [reply packages/data-designer-config/src/data_designer/config/base.py]: Addressed in 0bb974b8. `_get_docstring_summary` now checks against a `frozenset` of all standard Google-style section headers (`Args:`, `Returns:`, `Raises:`, `Note:`, `Example:`, etc.) instead of only `Attributes:`.

Re: the `__init__` note — `DataDesignerConfigBuilder.__init__` already has a docstring, so the summary field is already populated in the builder output.
